minetest/minetest
# Cascaded Shadow Maps

Upgrades the current shadow mapping implementation to cascaded shadow maps.

Benefits:
* Affine transformation for the shadow = simpler math, faster shader, less maintenance
* Fixes shadows for large objects/meshes such as e.g. in Little Lady or shadows for clouds in the future
* Smoother updates of shadows (every cascade updates at its own pace, cascade 0 updates every frame)
* Much less peter-panning
* Less VRAM consumption for similar shadow quality/scale
* Visible area covered with shadows in most cases even with fast camera movement

Other fixes:
* Removed flicker when mapblock is updated (e.g. flowing water)
* Removed flicker when camera offset changes (every 200 nodes)
* Added bilinear filter for SHADOW_FILTER == 1 for playable shadow quality with more performance

Notes:
* 3 cascades
* Cascade 0 is fixed small size in world and updates every frame, giving reasonable shadow quality close to the player
* Cascade 2 is updated every shadow_update_frames frame (up to 32) and cascade 1 somewhere in-between
* Entity shadows on cascades 0 and 1
* Colored shadows work
* Soft shadows work (may need some tuning)
* Shadow performance with client_mesh_chunk_size = 1 may be lower because of increased number of drawcalls covering larger area. Increasing shadow_update_frames should help spread out the updates.
* shadow_map_texture_size now controls resolution of each of 3 cascades (2 for entities) and should be divided by 2 when migrating

## To do

This PR is Ready for Review.

- [ ] Testing needed
- [ ] Tuning/defaults feedback needed

## How to test

1. Turn on the shadows
2. Enter a game that supports shadows
3. Check the shadows both at close and long distance, when flying etc.
